Unable to compile if the ProxyManager is in objective c++

### Bug Report
Due to our project needs, our proxy manager is in objective c++, this will cause compile error due to the template keyword that is used in SDLTemplateConfiguration.h.
This wasn't an issue in older version since SDLTemplateConfiguration class is added in 6.4.0

##### Reproduction Steps
1. Rename ProxyManager.m to ProxyManager.mm so it will be compiled as objective c++
2. Import SmartDeviceLink.h as documented
3. Compile error will occur.
Pods/Headers/Public/SmartDeviceLink/SDLTemplateConfiguration.h:40:46: error: expected identifier; 'template' is a keyword in Objective-C++

Pods/Headers/Public/SmartDeviceLink/SDLTemplateConfiguration.h:45:41: error: expected member name or ';' after declaration specifiers; 'template' is a keyword in Objective-C++
@property (strong, nonatomic) NSString *template;

Suggest to rename this variable to templateName instead of template

##### Expected Behavior
Project should compile fine.

##### Observed Behavior
Compoile failed

##### OS & Version Information
Xcode 11.2.1
SDL iOS SDK 6.4.0/6.4.1

--- a/SmartDeviceLink/SDLTemplateConfiguration.h
+++ b/SmartDeviceLink/SDLTemplateConfiguration.h
@@ -22,6 +22,7 @@ NS_ASSUME_NONNULL_BEGIN
  */
 - (instancetype)initWithPredefinedLayout:(SDLPredefinedLayout)predefinedLayout;
 
+// HAX: We are doing this because `template` is a C++ keyword and won't compile.
 #ifndef __cplusplus
 /**
  Init with the required values.

--- a/SmartDeviceLink/SDLTemplateConfiguration.h
+++ b/SmartDeviceLink/SDLTemplateConfiguration.h
@@ -9,7 +9,7 @@ NS_ASSUME_NONNULL_BEGIN
 
 /**
  Used to set an alternate template layout to a window.
- 
+
  @since SDL 6.0
  */
 @interface SDLTemplateConfiguration : SDLRPCStruct
@@ -22,9 +22,10 @@ NS_ASSUME_NONNULL_BEGIN
  */
 - (instancetype)initWithPredefinedLayout:(SDLPredefinedLayout)predefinedLayout;
 
+#ifndef __cplusplus
 /**
  Init with the required values.
- 
+
  @param template Predefined or dynamically created window template. Currently only predefined window template layouts are defined.
  */
 - (instancetype)initWithTemplate:(NSString *)template;
@@ -32,7 +33,7 @@ NS_ASSUME_NONNULL_BEGIN
 
 /**
  Convinience constructor with all the parameters.
- 
+
  @param template Predefined or dynamically created window template. Currently only predefined window template layouts are defined.
  @param dayColorScheme The color scheme to use when the head unit is in a light / day situation. If nil, the existing color scheme will be used.
  @param nightColorScheme The color scheme to use when the head unit is in a dark / night situation.
@@ -43,6 +44,31 @@ NS_ASSUME_NONNULL_BEGIN
  Predefined or dynamically created window template. Currently only predefined window template layouts are defined.
  */
 @property (strong, nonatomic) NSString *template;
+#endif
+
+#ifdef __cplusplus
+/**
+ Init with the required values.
+
+ @param templateName Predefined or dynamically created window template. Currently only predefined window template layouts are defined.
+ */
+- (instancetype)initWithTemplate:(NSString *)templateName;
+
+
+/**
+ Convinience constructor with all the parameters.
+
+ @param templateName Predefined or dynamically created window template. Currently only predefined window template layouts are defined.
+ @param dayColorScheme The color scheme to use when the head unit is in a light / day situation. If nil, the existing color scheme will be used.
+ @param nightColorScheme The color scheme to use when the head unit is in a dark / night situation.
+ */
+- (instancetype)initWithTemplate:(NSString *)templateName dayColorScheme:(nullable SDLTemplateColorScheme *)dayColorScheme nightColorScheme:(nullable SDLTemplateColorScheme *)nightColorScheme;
+
+/**
+ Predefined or dynamically created window template. Currently only predefined window template layouts are defined.
+ */
+@property (strong, nonatomic) NSString *templateName;
+#endif
 
 /**
  The color scheme to use when the head unit is in a light / day situation.

--- a/SmartDeviceLink/SDLTemplateConfiguration.m
+++ b/SmartDeviceLink/SDLTemplateConfiguration.m
@@ -1,60 +1,85 @@
 //
-//  SDLTemplateConfiguration.m
+//  SDLTemplateConfiguration.h
 //  SmartDeviceLink
 
-#import "SDLTemplateConfiguration.h"
+#import "SDLTemplateColorScheme.h"
+#import "SDLPredefinedLayout.h"
 
-#import "NSMutableDictionary+Store.h"
-#import "SDLRPCParameterNames.h"
+NS_ASSUME_NONNULL_BEGIN
 
-@implementation SDLTemplateConfiguration
+/**
+ Used to set an alternate template layout to a window.
+
+ @since SDL 6.0
+ */
+@interface SDLTemplateConfiguration : SDLRPCStruct
 
 
-- (instancetype)initWithPredefinedLayout:(SDLPredefinedLayout)predefinedLayout {
-    return [self initWithTemplate:predefinedLayout];
-}
+/**
+ Constructor with the required values.
 
-- (instancetype)initWithTemplate:(NSString *)template {
-    self = [super init];
-    if (!self) {
-        return nil;
-    }
-    self.template = template;
-    return self;
-}
+ @param predefinedLayout A template layout an app uses to display information. The broad details of the layout are defined, but the details depend on the IVI system. Used in SetDisplayLayout.
+ */
+- (instancetype)initWithPredefinedLayout:(SDLPredefinedLayout)predefinedLayout;
 
-- (instancetype)initWithTemplate:(NSString *)template dayColorScheme:(nullable SDLTemplateColorScheme *)dayColorScheme nightColorScheme:(nullable SDLTemplateColorScheme *)nightColorScheme {
-    self = [self initWithTemplate:template];
-    if (!self) {
-        return nil;
-    }
-    self.dayColorScheme = dayColorScheme;
-    self.nightColorScheme = nightColorScheme;
-    return self;
-}
+#ifndef __cplusplus
+/**
+ Init with the required values.
 
-- (void)setTemplate:(NSString *)template {
-    [self.store sdl_setObject:template forName:SDLRPCParameterNameTemplate];
-}
+ @param template Predefined or dynamically created window template. Currently only predefined window template layouts are defined.
+ */
+- (instancetype)initWithTemplate:(NSString *)template;
 
-- (NSString *)template {
-    return [self.store sdl_objectForName:SDLRPCParameterNameTemplate ofClass:NSString.class error:nil];
-}
 
-- (void)setDayColorScheme:(nullable SDLTemplateColorScheme *)dayColorScheme {
-    [self.store sdl_setObject:dayColorScheme forName:SDLRPCParameterNameDayColorScheme];
-}
+/**
+ Convinience constructor with all the parameters.
 
-- (nullable SDLTemplateColorScheme *)dayColorScheme {
-    return [self.store sdl_objectForName:SDLRPCParameterNameDayColorScheme ofClass:SDLTemplateColorScheme.class error:nil];
-}
+ @param template Predefined or dynamically created window template. Currently only predefined window template layouts are defined.
+ @param dayColorScheme The color scheme to use when the head unit is in a light / day situation. If nil, the existing color scheme will be used.
+ @param nightColorScheme The color scheme to use when the head unit is in a dark / night situation.
+ */
+- (instancetype)initWithTemplate:(NSString *)template dayColorScheme:(nullable SDLTemplateColorScheme *)dayColorScheme nightColorScheme:(nullable SDLTemplateColorScheme *)nightColorScheme;
 
-- (void)setNightColorScheme:(nullable SDLTemplateColorScheme *)nightColorScheme {
-    [self.store sdl_setObject:nightColorScheme forName:SDLRPCParameterNameNightColorScheme];
-}
+/**
+ Predefined or dynamically created window template. Currently only predefined window template layouts are defined.
+ */
+@property (strong, nonatomic) NSString *template;
+#endif
 
-- (nullable SDLTemplateColorScheme *)nightColorScheme {
-    return [self.store sdl_objectForName:SDLRPCParameterNameNightColorScheme ofClass:SDLTemplateColorScheme.class error:nil];
-}
+#ifdef __cplusplus
+/**
+ Init with the required values.
+
+ @param templateName Predefined or dynamically created window template. Currently only predefined window template layouts are defined.
+ */
+- (instancetype)initWithTemplate:(NSString *)templateName;
+
+
+/**
+ Convinience constructor with all the parameters.
+
+ @param templateName Predefined or dynamically created window template. Currently only predefined window template layouts are defined.
+ @param dayColorScheme The color scheme to use when the head unit is in a light / day situation. If nil, the existing color scheme will be used.
+ @param nightColorScheme The color scheme to use when the head unit is in a dark / night situation.
+ */
+- (instancetype)initWithTemplate:(NSString *)templateName dayColorScheme:(nullable SDLTemplateColorScheme *)dayColorScheme nightColorScheme:(nullable SDLTemplateColorScheme *)nightColorScheme;
+
+/**
+ Predefined or dynamically created window template. Currently only predefined window template layouts are defined.
+ */
+@property (strong, nonatomic) NSString *templateName;
+#endif
+
+/**
+ The color scheme to use when the head unit is in a light / day situation.
+ */
+@property (strong, nonatomic, nullable) SDLTemplateColorScheme *dayColorScheme;
+
+/**
+ The color scheme to use when the head unit is in a dark / night situation.
+ */
+@property (strong, nonatomic, nullable) SDLTemplateColorScheme *nightColorScheme;
 
 @end
+
+NS_ASSUME_NONNULL_END

--- a/SmartDeviceLink/SDLTemplateConfiguration.m
+++ b/SmartDeviceLink/SDLTemplateConfiguration.m
@@ -1,85 +1,91 @@
 //
-//  SDLTemplateConfiguration.h
+//  SDLTemplateConfiguration.m
 //  SmartDeviceLink
 
-#import "SDLTemplateColorScheme.h"
-#import "SDLPredefinedLayout.h"
+#import "SDLTemplateConfiguration.h"
 
-NS_ASSUME_NONNULL_BEGIN
+#import "NSMutableDictionary+Store.h"
+#import "SDLRPCParameterNames.h"
 
-/**
- Used to set an alternate template layout to a window.
-
- @since SDL 6.0
- */
-@interface SDLTemplateConfiguration : SDLRPCStruct
+@implementation SDLTemplateConfiguration
 
 
-/**
- Constructor with the required values.
-
- @param predefinedLayout A template layout an app uses to display information. The broad details of the layout are defined, but the details depend on the IVI system. Used in SetDisplayLayout.
- */
-- (instancetype)initWithPredefinedLayout:(SDLPredefinedLayout)predefinedLayout;
+- (instancetype)initWithPredefinedLayout:(SDLPredefinedLayout)predefinedLayout {
+    return [self initWithTemplate:predefinedLayout];
+}
 
 #ifndef __cplusplus
-/**
- Init with the required values.
+- (instancetype)initWithTemplate:(NSString *)template {
+    self = [super init];
+    if (!self) {
+        return nil;
+    }
+    self.template = template;
+    return self;
+}
 
- @param template Predefined or dynamically created window template. Currently only predefined window template layouts are defined.
- */
-- (instancetype)initWithTemplate:(NSString *)template;
+- (instancetype)initWithTemplate:(NSString *)template dayColorScheme:(nullable SDLTemplateColorScheme *)dayColorScheme nightColorScheme:(nullable SDLTemplateColorScheme *)nightColorScheme {
+    self = [self initWithTemplate:template];
+    if (!self) {
+        return nil;
+    }
+    self.dayColorScheme = dayColorScheme;
+    self.nightColorScheme = nightColorScheme;
+    return self;
+}
 
+- (void)setTemplate:(NSString *)template {
+    [self.store sdl_setObject:template forName:SDLRPCParameterNameTemplate];
+}
 
-/**
- Convinience constructor with all the parameters.
-
- @param template Predefined or dynamically created window template. Currently only predefined window template layouts are defined.
- @param dayColorScheme The color scheme to use when the head unit is in a light / day situation. If nil, the existing color scheme will be used.
- @param nightColorScheme The color scheme to use when the head unit is in a dark / night situation.
- */
-- (instancetype)initWithTemplate:(NSString *)template dayColorScheme:(nullable SDLTemplateColorScheme *)dayColorScheme nightColorScheme:(nullable SDLTemplateColorScheme *)nightColorScheme;
-
-/**
- Predefined or dynamically created window template. Currently only predefined window template layouts are defined.
- */
-@property (strong, nonatomic) NSString *template;
+- (NSString *)template {
+    return [self.store sdl_objectForName:SDLRPCParameterNameTemplate ofClass:NSString.class error:nil];
+}
 #endif
 
-#ifdef __cplusplus
-/**
- Init with the required values.
+#ifdef _cplusplus
+- (instancetype)initWithTemplate:(NSString *)templateName {
+    self = [super init];
+    if (!self) {
+        return nil;
+    }
+    self.templateName = templateName;
+    return self;
+}
 
- @param templateName Predefined or dynamically created window template. Currently only predefined window template layouts are defined.
- */
-- (instancetype)initWithTemplate:(NSString *)templateName;
+- (instancetype)initWithTemplate:(NSString *)templateName dayColorScheme:(nullable SDLTemplateColorScheme *)dayColorScheme nightColorScheme:(nullable SDLTemplateColorScheme *)nightColorScheme; {
+    self = [self initWithTemplate:templateName];
+    if (!self) {
+        return nil;
+    }
+    self.dayColorScheme = dayColorScheme;
+    self.nightColorScheme = nightColorScheme;
+    return self;
+}
 
+- (void)setTemplateName:(NSString *)templateName {
+    [self.store sdl_setObject:template forName:SDLRPCParameterNameTemplate];
+}
 
-/**
- Convinience constructor with all the parameters.
-
- @param templateName Predefined or dynamically created window template. Currently only predefined window template layouts are defined.
- @param dayColorScheme The color scheme to use when the head unit is in a light / day situation. If nil, the existing color scheme will be used.
- @param nightColorScheme The color scheme to use when the head unit is in a dark / night situation.
- */
-- (instancetype)initWithTemplate:(NSString *)templateName dayColorScheme:(nullable SDLTemplateColorScheme *)dayColorScheme nightColorScheme:(nullable SDLTemplateColorScheme *)nightColorScheme;
-
-/**
- Predefined or dynamically created window template. Currently only predefined window template layouts are defined.
- */
-@property (strong, nonatomic) NSString *templateName;
+- (NSString *)templateName {
+    return [self.store sdl_objectForName:SDLRPCParameterNameTemplate ofClass:NSString.class error:nil];
+}
 #endif
 
-/**
- The color scheme to use when the head unit is in a light / day situation.
- */
-@property (strong, nonatomic, nullable) SDLTemplateColorScheme *dayColorScheme;
+- (void)setDayColorScheme:(nullable SDLTemplateColorScheme *)dayColorScheme {
+    [self.store sdl_setObject:dayColorScheme forName:SDLRPCParameterNameDayColorScheme];
+}
 
-/**
- The color scheme to use when the head unit is in a dark / night situation.
- */
-@property (strong, nonatomic, nullable) SDLTemplateColorScheme *nightColorScheme;
+- (nullable SDLTemplateColorScheme *)dayColorScheme {
+    return [self.store sdl_objectForName:SDLRPCParameterNameDayColorScheme ofClass:SDLTemplateColorScheme.class error:nil];
+}
+
+- (void)setNightColorScheme:(nullable SDLTemplateColorScheme *)nightColorScheme {
+    [self.store sdl_setObject:nightColorScheme forName:SDLRPCParameterNameNightColorScheme];
+}
+
+- (nullable SDLTemplateColorScheme *)nightColorScheme {
+    return [self.store sdl_objectForName:SDLRPCParameterNameNightColorScheme ofClass:SDLTemplateColorScheme.class error:nil];
+}
 
 @end
-
-NS_ASSUME_NONNULL_END

